### PR TITLE
Provide a setting to control how fixed-length fields are truncated

### PR DESCRIPTION
**Pull request by Greybird**
Hello,

I would like to propose a new setting for fixed-length fields to control how an overflowing value is managed.

Basically, currently, the behavior is to discard characters at the end of the value.
I added a second option, which is to throw an exception when this occurs, as a way to ensure users can be warned if data was lost.
Of course, this could be the foundation to other strategies for handling this case.

Regards,

### CODE REVIEW

**Comment by Greybird** [comment]
Hi @MarcosMeli 

Do you think this PR could be integrated in the next version ?
Do you need further work to be done before it can be considered ? If so, would be happy to fill the gaps!

**Comment by MarcosMeli** [comment]
HI @Greybird  thanks for your PR !!

I really like idea, lets discuss some things here so we can move forward:

1. I like the idea to default to previous behavior, no breaking changes, I like the idea
2. We need to add a third option that is: KeepToEOL or something like
3. We need to check the attribute is added to the last field of the record
4. We need to add more tests to validate the behavior in record clases with more fields

what do you think ?
Thanks in advance

**Comment by Greybird** [comment]
Hi,

Thanks for your feedback!

For KeepToEOL, you mean have a behavior that states that the truncation does not occur on the last field of a record ? Wouldn't this be equal to having a field length of int.MaxValue ?
As the overflow occurs on the field level, it is quite strange to me to include a behavior depending on the position.
So I think I might be misinterpreting what you mean, can you please elaborate?

Thanks

**Comment by Greybird** [comment]
@marcomeli ping?

**Comment by Greybird** [comment]
Hello!
Any chance to have this PR merged ?
Regards!

**Comment by mcavigelli** [comment]
Salut Greybird.

I will look into it within the next two weeks.

Bon weekend, Matthias

File: FileHelpers.Tests/Tests/Common/OverflowModeTests.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        [Test]
+        public void Discard()
+        {
+            var engine = new FixedFileEngine<DiscardCustomer>();
+            var customer = new DiscardCustomer {mCustomerID = "0123456789ABC"};
+            var res = engine.WriteString(new[] {customer});
+
+            Check.That(res).IsEqualTo($"0123456789{Environment.NewLine}");
+        }
+
```
Comment (mcavigelli) [inline FileHelpers.Tests/Tests/Common/OverflowModeTests.cs:35]: Could you add another test, that just covers the edge case length=10, text.length = 10 and option Discard?

File: FileHelpers/Attributes/FieldFixedLengthAttribute.cs [hunk 2]
```diff
@@ -11,6 +12,9 @@ public sealed class FieldFixedLengthAttribute : FieldAttribute
         /// <summary>Length of this Fixed Length field.</summary>
         public int Length { get; private set; }
 
+        /// <summary>Overflow behavior for this Fixed Length field</summary>
+        public OverflowMode OverflowMode { get; set; } = OverflowMode.DiscardEnd;
```
Comment (mcavigelli) [inline FileHelpers/Attributes/FieldFixedLengthAttribute.cs:16]: I think here the good documentation should also contain the default behaviour (Discard).

File: FileHelpers/Fields/FixedLengthField.cs [hunk 3]
```diff
@@ -27,6 +29,11 @@ public sealed class FixedLengthField
         /// </summary>
         internal FixedMode FixedMode { get; set; }
 
+        /// <summary>
+        /// How an overflowing field value is handled
+        /// </summary>
+        internal OverflowMode OverflowMode { get; set; }
```
Comment (mcavigelli) [inline FileHelpers/Fields/FixedLengthField.cs:35]: I think private visibility is sufficient.

File: FileHelpers/Fields/FixedLengthField.cs [hunk 4]
```diff
@@ -123,7 +132,9 @@ internal override void CreateFieldString(StringBuilder sb, object fieldValue, bo
 
             // Discard longer field values
             if (field.Length > FieldLength)
-                field = field.Substring(0, FieldLength);
+            {
```
Comment (mcavigelli) [inline FileHelpers/Fields/FixedLengthField.cs:135]: The logic for text is split here and in the method GetValueForOverflowingField.
I think it would be better, just to call the method anyway and handle the whole logic there.

**Review by mcavigelli — CHANGES_REQUESTED** [review]
Hi there

Thank you for your enhancement of the FileHelpers library.
I have added some comments. The changes I suggest are of minor size.

Matthias

File: FileHelpers.Tests/Tests/Common/OverflowModeTests.cs (hunk 1 above)
Comment (Greybird) [reply FileHelpers.Tests/Tests/Common/OverflowModeTests.cs:37]: Sure, added some use cases with 9/10/11 too.

**Comment by Greybird** [comment]
Hi @mcavigelli,

Thanks for your review!
I took your comments into account, and merged master into this branch.
Do not hesitate if you find other enhancements!

Arnaud

**Comment by mcavigelli** [comment]
Thank you for your quick reply.
I will look into it this week.

Matthias

File: FileHelpers.Tests/Tests/Common/OverflowModeTests.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        [FixedLengthRecord]
+        public class ErrorCustomer
+        {
+            [FieldFixedLength(10, OverflowMode = OverflowMode.Error)]
+            public string mCustomerID;
+        }
+
+        [TestCase("0123456789ABC", "0123456789")]
```
Comment (mcavigelli) [inline FileHelpers.Tests/Tests/Common/OverflowModeTests.cs:25]: Thank you for your extensive tests.

File: FileHelpers/Attributes/FieldNullValueAttribute.cs [hunk 6]
```diff
@@ -23,7 +23,7 @@ public sealed class FieldNullValueAttribute : Attribute
 
         /// <summary>
         /// Defines the default in event of a null value.
-        /// Object must be of teh correct type
+        /// Object must be of the correct type
```
Comment (mcavigelli) [inline FileHelpers/Attributes/FieldNullValueAttribute.cs:26]: Merci.

**Review by mcavigelli — APPROVED** [review]
Thanks again for your great work.
And thank you for your patience.

Matthias
